src/server.py

```diff
@@ -1,45 +1,151 @@
     server.listen(1)
     print('starting')
     try:
         while True:
             conn, addr = server.accept()
-            buffer_length = 8
-            message_complete = False
-            full_string = b""
-            while not message_complete:
-                part = conn.recv(buffer_length)
-                full_string = full_string + part
-                if len(part) < buffer_length:
-                    message_complete = True
-            print(full_string.decode('utf8'))
-            full_string += b"\r\n\r\n" 
-            conn.sendall(full_string)
-            conn.sendall(response_ok())
+            try:
+                buffer_length = 8
+                message_complete = False
+                full_string = b""
+                while not message_complete:
+                    part = conn.recv(buffer_length)
+                    full_string = full_string + part
+                    if len(part) < buffer_length:
+                        full_string = full_string.decode('utf-8')
+                        message_complete = True
+                # print('full_string.decode', full_string)
+                try:
+                    uri = parse_request(full_string)
+                    body_tuple = resolve_uri(uri)
+                    if body_tuple:
+                        response = response_ok(body_tuple)
+                    else:
+                        response = response_error('404 Not Found')
+                    response = response_ok(body)
+                except SyntaxError:
+                    response = response_error('404 Page Not Found')
+                except NameError:
+                    response = response_error('405 Method Not Allowed')
+                except TypeError:
+                    response = response_error('505 HTTP Version Not Supported')
+                except ValueError:
+                    raise HTTPException(
+                        HTTP_BAD_REQUEST,
+                        "No 'Host :' in header")
+            except SystemError:
+                response = response_error()
+            # full_string += b"\r\n\r\n"
+            # conn.sendall(full_string)
+            conn.sendall(response.encode('utf-8'))
             conn.close()
     except KeyboardInterrupt:
         server.close()
 
 
-def response_ok():
-    """Return HTTP 200 OK message for when the connection is ok."""
-    return (b"HTTP/1.1 200 OK\r\n\r\n"
-            b"Success")
+def response_ok(body_tuple):
+    """Return 200 ok."""
+    first = 'HTTP/1.1 200 OK'
+    second_line = 'Content-Type:' + body_tuple[1] + '; charset=utf-8'
+    date = 'Date: ' + email.utils.formatdate(usegmt=True)
+    header_break = ''
+    body = body_tuple
+    fourth_line = 'Content-Length: {}'.format(len(body))
+    string_list = [first, second_line, date, fourth_line, header_break, body]
+    string_list = '\r\n'.join(string_list) + '\r\n'
+    string_list = string_list.encode('utf-8') + body
+    return string_list
```

Comment:
by my reading, we are now expecting the return value of `response_ok` to be bytes.  that's great!  Remember that when choosing to encode/decode the data that comes from here.